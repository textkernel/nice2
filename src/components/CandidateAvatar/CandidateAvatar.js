import React from 'react';
import PropTypes from 'prop-types';
import bem from 'bem';
import styles from './CandidateAvatar.scss';

const { block, elem } = bem({
    name: 'CandidateAvatar',
    classnames: styles,
    propsToMods: ['context']
});

const context = percentage => {
    if (percentage <= 33) {
        return 'bad';
    }
    if (percentage <= 66) {
        return 'warning';
    }
    return 'good';
};

const CandidateAvatar = props => {
    const { imageUrl, matchPercentage, showPercentageOnHover, size, ...rest } = props;

    const constrainedSize = Math.round(Math.max(32, size));
    const fixedSize = constrainedSize % 2 === 0 ? constrainedSize : constrainedSize + 1; // force even number
    const isSmall = fixedSize < 60;
    const strokeWidth = isSmall ? 2 : 4;
    const radiusCorrection = isSmall ? 3 : 2;
    const radius = fixedSize / 2 - radiusCorrection;
    const percentage = Math.max(0, Math.min(100, matchPercentage));
    const circumference = 2 * radius * Math.PI;
    const strokeDasharray = `${(percentage * circumference) / 100} 999`;

    return (
        <div
            {...rest}
            {...block(props)}
            style={{
                width: fixedSize,
                height: fixedSize
            }}
        >
            <div
                {...elem('image', props)}
                style={{
                    backgroundImage: `url(${imageUrl})`
                }}
            >
                {!!Number.isInteger(matchPercentage) &&
                    !!showPercentageOnHover && (
                        <svg {...elem('percentage', props)} viewBox="0 0 50 50">
                            <text
                                {...elem('percentage-value', props)}
                                x="50%"
                                y="50%"
                                textAnchor="middle"
                                dominantBaseline="central"
                            >
                                {`${percentage}%`}
                            </text>
                        </svg>
                    )}
            </div>
            <svg
                {...elem('ring', props)}
                style={{
                    width: fixedSize,
                    height: fixedSize
                }}
            >
                {!!Number.isInteger(matchPercentage) && (
                    <circle
                        r={radius}
                        cx={fixedSize / 2}
                        cy={fixedSize / 2}
                        strokeWidth={strokeWidth}
                        {...elem('circle', {
                            ...props,
                            context: context(percentage)
                        })}
                        style={{
                            strokeDasharray
                        }}
                    />
                )}
            </svg>
        </div>
    );
};

CandidateAvatar.displayName = 'CandidateAvatar';

CandidateAvatar.propTypes = {
    /** Path to the avatar image resource */
    imageUrl: PropTypes.string,
    /** The match percentage, ranging from 0 (bad) - 100 (good) */
    matchPercentage: PropTypes.number,
    /** Whether to show the match percentage when hovering the image */
    showPercentageOnHover: PropTypes.bool,
    /** Avatar size (1:1 aspect ratio), should be even number */
    size: PropTypes.number
};

CandidateAvatar.defaultProps = {
    imageUrl: null,
    matchPercentage: null,
    showPercentageOnHover: false,
    size: 72
};

export default CandidateAvatar;
