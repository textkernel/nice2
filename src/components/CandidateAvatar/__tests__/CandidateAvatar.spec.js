import React from 'react';
import toJson from 'enzyme-to-json';
import CandidateAvatar from '../CandidateAvatar';

describe('<CandidateAvatar> that renders a candidate profile image with match indication', () => {
    it('should render a default avatar', () => {
        const wrapper = shallow(<CandidateAvatar />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render a custom sized avatar with good match percentage', () => {
        // Try odd size to see if it is forced to be even
        const wrapper = shallow(<CandidateAvatar size={127} matchPercentage={100} />);
        expect(wrapper.props().style).toEqual({
            height: 128,
            width: 128,
        });
        expect(wrapper.find('svg').props().style).toEqual({
            height: 128,
            width: 128,
        });
        expect(wrapper.find('circle').props().strokeWidth).toBe(4);
        expect(wrapper.find('circle').props().cx).toBe(64);
        expect(wrapper.find('circle').props().cy).toBe(64);
    });

    it('should render an avatar with average match percentage', () => {
        const wrapper = shallow(<CandidateAvatar size={128} matchPercentage={34} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render an avatar with bad match percentage and avatar', () => {
        const wrapper = shallow(
            <CandidateAvatar
                imageUrl="/candidate.jpg"
                size={128}
                matchPercentage={33}
                showPercentageOnHover
            />
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should change avatar image', () => {
        const wrapper = shallow(<CandidateAvatar imageUrl="/candidate.jpg" />);
        expect(wrapper.find('.CandidateAvatar__image').props().style).toEqual({
            backgroundImage: 'url(/candidate.jpg)',
        });
    });

    it('should add classes when props are changed', () => {
        const wrapper = shallow(
            <CandidateAvatar showPercentageOnHover matchPercentage={10} size={58} />
        );
        expect(wrapper.find('.CandidateAvatar__percentage')).toHaveLength(1);
        expect(wrapper.find('.CandidateAvatar__percentage').text()).toBe('10%');
        expect(wrapper.find('circle').props().strokeWidth).toBe(2);
    });
});
