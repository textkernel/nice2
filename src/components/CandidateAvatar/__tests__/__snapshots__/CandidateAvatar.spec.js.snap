// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CandidateAvatar> that renders a candidate profile image with match indication should render a default avatar 1`] = `
<div
  className="CandidateAvatar"
  style={
    Object {
      "height": 72,
      "width": 72,
    }
  }
>
  <div
    className="CandidateAvatar__image"
    style={
      Object {
        "backgroundImage": "url(null)",
      }
    }
  />
  <svg
    className="CandidateAvatar__ring"
    style={
      Object {
        "height": 72,
        "width": 72,
      }
    }
  />
</div>
`;

exports[`<CandidateAvatar> that renders a candidate profile image with match indication should render an avatar with average match percentage 1`] = `
<div
  className="CandidateAvatar"
  style={
    Object {
      "height": 128,
      "width": 128,
    }
  }
>
  <div
    className="CandidateAvatar__image"
    style={
      Object {
        "backgroundImage": "url(null)",
      }
    }
  />
  <svg
    className="CandidateAvatar__ring"
    style={
      Object {
        "height": 128,
        "width": 128,
      }
    }
  >
    <circle
      className="CandidateAvatar__circle CandidateAvatar__circle--context CandidateAvatar__circle--context_warning"
      cx={64}
      cy={64}
      r={62}
      strokeWidth={4}
      style={
        Object {
          "strokeDasharray": "132.44954627534568 999",
        }
      }
    />
  </svg>
</div>
`;

exports[`<CandidateAvatar> that renders a candidate profile image with match indication should render an avatar with bad match percentage 1`] = `
<div
  className="CandidateAvatar"
  style={
    Object {
      "height": 128,
      "width": 128,
    }
  }
>
  <div
    className="CandidateAvatar__image"
    style={
      Object {
        "backgroundImage": "url(null)",
      }
    }
  />
  <svg
    className="CandidateAvatar__ring"
    style={
      Object {
        "height": 128,
        "width": 128,
      }
    }
  >
    <circle
      className="CandidateAvatar__circle CandidateAvatar__circle--context CandidateAvatar__circle--context_bad"
      cx={64}
      cy={64}
      r={62}
      strokeWidth={4}
      style={
        Object {
          "strokeDasharray": "128.5539713848943 999",
        }
      }
    />
  </svg>
</div>
`;
